File: Linear_Probing.py
TABLE_SIZE = 10

class HashTable:
    def __init__(self):
        self.table = [-1] * TABLE_SIZE

    def hash_function(self, key):
        return key % TABLE_SIZE

    def insert(self, key):
        index = self.hash_function(key)
        original_index = index
        while self.table[index] != -1:
            index = (index + 1) % TABLE_SIZE
            if index == original_index:
                print(f"Hash table is full! Cannot insert {key}")
                return
        self.table[index] = key

    def display(self):
        print("\nHash Table:")
        for i in range(TABLE_SIZE):
            print(f"{i} => {self.table[i]}")

if __name__ == "__main__":
    ht = HashTable()

    n = int(input("Enter the number of elements to insert: "))

    for i in range(n):
        key = int(input(f"Enter key {i + 1}: "))
        ht.insert(key)

    ht.display()

File: Linear_Probing.c
#include <stdio.h>
#define TABLE_SIZE 10

int hashTable[TABLE_SIZE];

void initialize() {
    for (int i = 0; i < TABLE_SIZE; i++) {
        hashTable[i] = -1; // -1 indicates empty slot
    }
}

int hashFunction(int key) {
    return key % TABLE_SIZE;
}

void insert(int key) {
    int index = hashFunction(key);
    int originalIndex = index;
    while (hashTable[index] != -1) {
        index = (index + 1) % TABLE_SIZE;
        if (index == originalIndex) {
            printf("Hash table is full! Cannot insert %d\n", key);
            return;
        }
    }
    hashTable[index] = key;
}

void display() {
    printf("\nHash Table:\n");
    for (int i = 0; i < TABLE_SIZE; i++) {
        printf("%d => %d\n", i, hashTable[i]);
    }
}

int main() {
    int n, key;
    initialize();

    printf("Enter the number of elements to insert: ");
    scanf("%d", &n);

    for (int i = 0; i < n; i++) {
        printf("Enter key %d: ", i + 1);
        scanf("%d", &key);
        insert(key);
    }

    display();

    return 0;
}
